lib/fetch/util.js

```diff
@@ -297,6 +298,101 @@ function tryUpgradeRequestToAPotentiallyTrustworthyURL (request) {
   // TODO
 }
 
+function createDeferredPromise () {
+  let res
+  let rej
+  const promise = new Promise((resolve, reject) => {
+    res = resolve
+    rej = reject
+  })
+
+  return { promise, resolve: res, reject: rej }
+}
+
+// https://streams.spec.whatwg.org/#readablestreamdefaultreader-read-all-bytes
+/** @param {ReadableStreamDefaultReader} reader */
+async function ReadableStreamDefaultReaderReadAllBytes (reader) {
+  // 1. Let promise be a new promise.
+  const promise = createDeferredPromise()
+
+  // 2. Let bytes be an empty byte sequence.
+  const bytes = []
+
+  // 3. Read-loop given reader, bytes, and promise.
+  const readLoop = async () => {
+    const { value: chunk, done } = await reader.read()
+
+    // close steps - Resolve promise with bytes
+    if (done) {
+      promise.resolve(bytes)
+      return
+    }
+
+    // 1. If chunk is not a Uint8Array object, reject promise with a TypeError and abort these steps.
+    if (!isUint8Array(chunk)) {
+      // error steps, given e - Reject promise with e.
+      promise.reject(new TypeError('invalid chunk type received'))
+      return
+    }
+
+    // 2. Append the bytes represented by chunk to bytes.
+    bytes.push(chunk)
+
+    // 3. Read-loop given reader, bytes, and promise.
+    // This recursion could potentially cause a stack overflow if implemented directly. Implementations will need to mitigate this, 
+    // e.g. by using a non-recursive variant of this algorithm, or queuing a microtask, or using a more direct method of
+    // byte-reading as noted below. 
+    await new Promise(res => queueMicrotask(res))
+    return readLoop()
+  }
+
+  await readLoop()
+
+  return promise
+}
+
+// https://fetch.spec.whatwg.org/#fully-reading-body-as-promise
+async function fullyReadBodyAsPromise (body) {
+  // 1. Let reader be the result of getting a reader for body’s stream. If that threw an exception,
+  //    then return a promise rejected with that exception.
+  const reader = body.stream.getReader()
+
+  // 2. Return the result of reading all bytes from reader.
+  return await ReadableStreamDefaultReaderReadAllBytes(reader)
+}
+
+// https://fetch.spec.whatwg.org/#body-fully-read
+function fullyReadBody (
+  // body body
+  body,
+  // given an algorithm processBody
+  processBody,
+  // an algorithm processBodyError,
+  processBodyError,
+  // and an optional null, parallel queue, or global object taskDestination (default null)
+  taskDestination = null
+) {
+  // 1. If taskDestination is null, then set taskDestination to the result of starting a new parallel queue.
+  if (taskDestination == null) {
+    // TODO
```

Comment:
This must be filled

Reply:
Why? It's not fully spec compliant per se but I don't think it makes much of a difference for the end user.

Reply:
Then we should just remove the if.

Reply:
task destinations aren't used elsewhere in the code and I didn't quite understand the spec's wording for them and the library, being in js, isn't really needed here to my understanding.

Reply:
https://html.spec.whatwg.org/multipage/infrastructure.html#parallel-queue

My issue with implementing a parallel queue is that it seemed quite... useless and the wording was a little strange. If we queue algorithms (functions) then we're just running them back to back, which works regardless in this library. Plus, the spec doesn't note *when* to start dequeuing so my understanding is that it's instant.

Reply:
The if has been removed in 09e0e798547ca4760363368c10c2fcdfb8f6fc74. Parallel queues are probably better left for another pr :)